### DISCUSSION TITLE
Detect a pattern of two consequtive lines, anywhere.

### DISCUSSION THREAD

**Discussion by amomchilov** [rubocop/rubocop#11937] [Q&A] [answered]
I have a [cop](https://github.com/Shopify/rubocop-sorbet/pull/162) that aims to detect 2 consecutive (ignoring whitespace) lines that have this pattern:

```ruby
def foo
  @foo = T.let(@foo, T.nilable(Foo))
  @foo ||= Foo.new
end
```

The current implementation works, but it only works if these two lines are the first thing in the method body. E.g. this won't be caught:

```ruby
def foo
  some
  other
  code
  @foo = T.let(@foo, T.nilable(Foo))
  @foo ||= Foo.new
end
```

What would be the right way to match these two lines, and replace them without effecting other lines before/after them?

🧵 https://github.com/Shopify/rubocop-sorbet/pull/162#discussion_r1213644528

**Comment by dvandersluis** [comment, marked as answer]
I briefly looked at your PR. You're looking for a `begin` node, which is going to capture the entire body of the method in your example. As you discussed in the PR, you would need use `...` to make the matcher match when there is extra code (likely before and after the nodes you're looking for), but this will still be matching the entire `begin` node.

You can use [capturing](https://docs.rubocop.org/rubocop-ast/node_pattern.html#for-captures) within your pattern to return the exact child nodes within the `begin` that you are concerned about, and pass those into your correction rather than the parent `begin` node.

**Reply by dvandersluis** [reply]
Here's a [node pattern example](https://nodepattern.herokuapp.com/?p=(begin%0A%20%20...%0A%20%20%24(ivasgn%20%24_ivar%0A%20%20%20%20(send%0A%20%20%20%20%20%20(const%20nil%3F%20%3AT)%20%3Alet%0A%20%20%20%20%20%20(ivar%20_ivar)%0A%20%20%20%20%20%20(send%20%0A%20%20%20%20%20%20%20%20(const%20nil%3F%20%3AT)%20%3Anilable%20%24_ivar_type)))%0A%20%20%24(or-asgn%0A%20%20%20%20(ivasgn%20_ivar)%0A%20%20%20%20%24_initialization_expr)%0A%20%20...%0A%20)%0A&ruby=def%20foo%0A%20%20some%0A%20%20other%0A%20%20code%0A%20%20%40foo%20%3D%20T.let(%40foo%2C%20T.nilable(Foo))%0A%20%20%40foo%20%7C%7C%3D%20Foo.new%0Aend). The `ivasgn` node is capture 1 and the `or_asgn` node is capture 4.

**Reply by amomchilov** [reply]
> you would need use ... to make the matcher match when there is extra code ... but this will still be matching the entire begin node.

Indeed, that was exactly the issue I was having (then auto-correction would replace _all_ the lines, and clobber that extra code)

> You can use [capturing](https://docs.rubocop.org/rubocop-ast/node_pattern.html#for-captures) within your pattern to return the exact child nodes within the begin that you are concerned about, and pass those into your correction rather than the parent begin node.

And that's the solution!

I forgot that you can capture an entire subtree... which is funny, because my current node-matcher already captures an entire subtree 😅

**Reply by amomchilov** [reply]
Hey Daniel, I got pretty close! Here's my attempt: https://github.com/Shopify/rubocop-sorbet/commit/cb802c97e50a034c042e61db219854a44d4d89c1

There's just one issue with this, which is the surrounding indentation around the nodes:

```diff
sig { returns(T::Hash[Symbol, T.untyped]) }
def client_info_hash
  @client_info_hash ||= T.let(client_info.to_hash, T.nilable(T::Hash[Symbol, T.untyped]))
+  
end
```

Previously, I was replacing the entire `begin` node as a whole. I would fish out the relevant parts, and build up the replacement without any excess whitespace.

I can't replace the entire `begin` node anymore (obviously, since that'll clobber those previous unrelated lines, that was the whole issue). Now that I'm capturing the two child nodes directly, I try to replace the first line with the correction, and remove the second line. Since these captures don't capture any trailing/leading white space, it survives the replacement+removal of the subnodes.

I can fidget around with manually stripping the whitespace, but it feels like I'm missing something there.

Got any suggestions?

**Reply by dvandersluis** [reply]
`range_by_whole_lines` from [`RangeHelp`](https://github.com/rubocop/rubocop/blob/master/lib/rubocop/cop/mixin/range_help.rb#L73) will probably help you here. You should be able to expand your range to the entire line including the final new line, and give that to the corrector to remove.

**Reply by amomchilov** [reply]
That got me pretty close! It got rid of the trailing line, but I still needed to do some manual work to remove the blank lines between my two lines of interest.

What do you think?

[Shopify/rubocop-sorbet@`4b51532` (#162)](https://github.com/Shopify/rubocop-sorbet/pull/162/commits/4b5153205bc1830ac027f4f2cdad3d179305fc9b#diff-c8ec9207ecefc0c8e80d6667311d9ddb51b5f46702b15588b9fb15b9e976a6b1R112-R119)
